### PR TITLE
Remove unstable APIs for /hierarchy.

### LINKED ISSUE #10946: Stabilize support for MSC2946: spaces summary
Once [MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946) lands we should (at some point) switch to a stable endpoint and make some updates. There's few bits to this that we need to do:

## Phase 1

* [x] Support to the stable endpoint.
* [x] ~~Require `limit` to be > 0.~~ (This has always been correct.)
* [x] Ensure the endpoint is ratelimited.

## Phase 2

* [x] Remove the fallback logic of `get_room_hierarchy` to request the `/spaces` endpoint over federation.
* [x] Remove the `/spaces` client and federation endpoints (and corresponding code in the handlers).
* [x] Update the API responses to remove `allowed_spaces` (this was replaced with `allowed_room_ids`).
* [x] Clean-up code around max-children, which is not used by the `/hierarchy` endpoint.
* [x] Remove `room_id` from the serialized `children_state`, it is redundant and removed from the API response.

## Phase 3

* [x] Remove unstable `/hierarchy` client and federation endpoints.
* [x] Remove the typo `join_rules` field from the response.

### PR DESCRIPTION
**Pull request by clokep**
Finish-up stabilization of MSC2946:

* Removes the unstable endpoint.
* Removes a typo in one of the response fields.
* Minor test clean-ups.

Fixes #10946

### CODE REVIEW

File: docs/workers.md [hunk 1]
```diff
@@ -205,7 +205,7 @@ information.
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/context/.*$
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/members$
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/state$
-    ^/_matrix/client/(v1|unstable/org.matrix.msc2946)/rooms/.*/hierarchy$
+    ^/_matrix/client/v1/rooms/.*/hierarchy$
```
Comment (erikjohnston) [inline docs/workers.md:208]: Shouldn't this be `v3`? Surely its not `v1`?

Comment (clokep) [reply docs/workers.md:208]: It is a new endpoint, so it was added as v1. The only endpoints that became v3 were the ones which existed as r0.

See https://github.com/matrix-org/synapse/blob/d4713d3e335b21d12284ddd8ebd00e38abcfd521/synapse/rest/client/room.py#L1196-L1201

File: docs/workers.md [hunk 2]
```diff
@@ -205,8 +205,8 @@ information.
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/context/.*$
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/members$
     ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/state$
-    ^/_matrix/client/(v1|unstable/org.matrix.msc2946)/rooms/.*/hierarchy$
-    ^/_matrix/client/(v1|unstable/org.matrix.msc2716)/rooms/.*/batch_send$
+    ^/_matrix/client/v1/rooms/.*/hierarchy$
+    ^/_matrix/client/unstable/org.matrix.msc2716/rooms/.*/batch_send$
```
Comment (clokep) [inline docs/workers.md:209]: This was wrong, see https://github.com/matrix-org/synapse/pull/12868#discussion_r881542588

File: docs/workers.md (hunk 1 above)
Comment (erikjohnston) [reply docs/workers.md:208]: Hmm, I forgot that that was how we decided to do it

**Review by erikjohnston — APPROVED** [review]

File: docs/workers.md (hunk 1 above)
Comment (clokep) [reply docs/workers.md:208]: It seems to confuse a lot of people. 😄

**Pull request merged**